### PR TITLE
Fix '*' normalization - make it round trippable

### LINKED ISSUE #6697: Wrong package dependency version If package dependency version is set to '*'
_From @atanasa on March 19, 2018 15:28_

If package dependency version is set to '*', the version in the package spec is set to the version of the package instead of the dependency

**Steps to reproduce:**

1. Build the following project:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
    <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
    <Version>1.2.3</Version>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="EntityFramework" Version="*" />
  </ItemGroup>

</Project>
```

Where "EntityFramework" can be any other package (probably that is not part of the standard library).

2. run `dotnet build`

**Expected output:**
A package of version 1.2.3 with a dependency to the package EntityFramework with it latest version (e.g. 6.2.0).

**Actual output:**
A package of version 1.2.3 with a dependency to the package EntityFramework with version 1.2.3:

```xml
<?xml version="1.0" encoding="utf-8"?>
<package xmlns="http://schemas.microsoft.com/packaging/2012/06/nuspec.xsd">
  <metadata>
    <id>ClassLibrary1</id>
    <version>1.2.3</version>
    <authors>ClassLibrary1</authors>
    <owners>ClassLibrary1</owners>
    <requireLicenseAcceptance>false</requireLicenseAcceptance>
    <description>Package Description</description>
    <dependencies>
      <group targetFramework=".NETStandard2.0">
        <dependency id="EntityFramework" version="1.2.3" exclude="Build,Analyzers" />
      </group>
    </dependencies>
  </metadata>
  <files>
    <file src="...\ReproPackDepVersion\ClassLibrary1\bin\Debug\netstandard2.0\ClassLibrary1.dll" target="lib\netstandard2.0\ClassLibrary1.dll" />
  </files>
</package>
```

`dotnet -info`:
```
.NET Command Line Tools (2.1.101)

Product Information:
 Version:            2.1.101
 Commit SHA-1 hash:  6c22303bf0

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.15063
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.1.101\

Microsoft .NET Core Shared Framework Host

  Version  : 2.0.6
  Build    : 74b1c703813c8910df5b96f304b0f2b78cdf194d
```

_Copied from original issue: dotnet/sdk#2062_

### LINKED ISSUE #8073: Lock file is not honored in "*" scenarios
Sample repo: https://github.com/NuGetPlay/TestLockFile

## Steps:
1. New project with `RestorePackagesWithLockFile` set
2. Take a dependency on a package with ` * `
3. Restore project
   * The lock file will be generated with the highest version (say **1.0.0**) present on the package source(s).
4. Add a newer version of the package (say **1.1.0**) to the repository (preferably local to avoid caching issues).
5. Delete obj, bin folders. 
6. Restore again, OR
7. Restore with `--locked-mode`

## Expected
* For 6, The restore should complete without modifying the lock file. Version **1.0.0** should be restored.
* For 7, the same should happen as 6. Restore **should not** fail.

## Actual
* For 6,  restore overrites the lock file, bringing in the version **1.1.0**
* For 7, restore **fails**.

### LINKED ISSUE #8432: NuGet.exe does not resolve to the latest version of a package when using * in PackageReference (MSBuild/Dotnet/VS restore do)
## Details about Problem

Please note this is the same issue described in #6640. But please do check the repro steps as it doesn't actually work. Unsure how it worked then.

```
Issue is if packagereference is set to "*" then it resolves to lowest available version and not highest one. In #6640 the discussion seems to suggest that highest version should be picked up.

What is the recommended method to get packagereference to resolve to highest available version?
This is a very useful feature, say if we are doing integration testing with latest available package.
```

NuGet product used (NuGet.exe | VS UI | Package Manager Console | dotnet.exe):
`NuGet.exe`

NuGet version (x.x.x.xxx):
`5.0.2.5988`

dotnet.exe --version (if appropriate):

VS version (if appropriate):

OS version (i.e. win10 v1607 (14393.321)):
`Win 10 Pro 1803`

Worked before? If so, with which NuGet version:
It has been suggested in #6640 that it used to work. But it has never worked for me.

## Detailed repro steps so we can see the same problem

1. Create an empty folder

2. Add test.csproj using the xml below

```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net45</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Selenium.WebDriver" Version="*"/>
  </ItemGroup>
</Project>
```

3. To start from a clean state, from a prompt run
`nuget locals all -clear`

4. Do a nuget restore with detailed output
`nuget restore -Verbosity detailed .\test.csproj`

See that Selenium.WebDriver has resolved to 2.2.0 and not the latest available package version on nuget.org.
`Acquired lock for the installation of Selenium.WebDriver 2.2.0`

What is the recommended mechanism to be able to resolve to latest available package version using package reference?

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/8432
Fixes: https://github.com/NuGet/Home/issues/8073
Fixes: https://github.com/NuGet/Home/issues/6697
Related: https://github.com/NuGet/Home/issues/8472

Regression: Yes/No  
* Last working version:   None
* How are we preventing it in future:   Tests

## Fix

Details: Read spec at: https://github.com/NuGet/Client.Engineering/pull/105

Make `*` round-trippable and obsolete ambiguous  statics in the VersionRangeFactory
This effectively changes `*` from round tripping to `(, )`, which means any available, take lowest to 
`[*, )` which conveys the select highest stable more accurately. 

+@joelverhagen as someone with lots of understanding of versions and normalization :) 

## Testing/Validation

Tests Added: Yes
Reason for not adding tests:  
Validation:

### CODE REVIEW

**Comment by nkolev92** [comment]
@NuGet/nuget-client 👀

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnresolvedMessages.cs [hunk 1]
```diff
@@ -264,9 +264,11 @@ internal static NuGetVersion GetBestMatch(SortedSet<NuGetVersion> versions, Vers
                 }
             }
 
-            if (!range.IsFloating || range.HasLowerBound || range.HasUpperBound)
+            // Take the lowest version higher than the pivot if one exists,
+            // unless the version range is floating without an upper bound
```
Comment (donnie-msft) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnresolvedMessages.cs]: Am I reading this wrong? The comment says "unless" (aka, if there's NOT) an upper bound. But the code says "floating AND hasUpperBound".

**Review by donnie-msft — APPROVED** [review]

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnresolvedMessages.cs (hunk 1 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnresolvedMessages.cs]: I think the comment is wrong. 
Basically if there's an upper bound we take the lowest version above the bound. 

So I think I'll change it to: 

> Take the lowest version higher than the pivot if one exists or for ranges that are floating with an upper bound

Any suggestions? 

The longer explanation is: 

* Floatless ranges - If there's a pivot/lower or higher bound, take the first above that pivot.
* Floating ranges - by definition they need the latest version, so unless there's an upper bound always show the highest version possible.

Comment (donnie-msft) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnresolvedMessages.cs]: Your sentence is concise, but a little hard to follow. I prefer more readable code and details like:
```
//Floatless ranges - If there's a pivot/lower or higher bound, take the first above that pivot.
bool floatlessRangeHasBounds = (!range.IsFloating && (range.HasLowerBound || range.HasUpperBound));

//Floating ranges - by definition they need the latest version, so unless there's an upper bound always show the highest version possible.
bool floatingRangeHasUpperBound = range.IsFloating && range.HasUpperBound;

if (floatlessRangeHasBounds || floatingRangeHasUpperBound)
{
     bestMatch = versions.Where(e => e >= ideal).FirstOrDefault();
}
```

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnresolvedMessages.cs]: Sure, I can do that.

Comment (joelverhagen) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnresolvedMessages.cs]: Having a truth table with common cases as rows and columns of these observed properties could help the reader (code comment)

File: test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionRangeTests.cs [hunk 2]
```diff
@@ -146,6 +146,32 @@ public void VersionRange_MetadataIsIgnored_FormatRemovesMetadata()
             Assert.Equal("[1.0.0, 2.0.0]", bothMetadata.ToLegacyString());
         }
 
+        [Fact]
+        public void VersionRange_FloatAllStable_ReturnsCorrectPrints()
```
Comment (joelverhagen) [inline test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionRangeTests.cs:179]: Worth having a true round trip test? `new` -> `ToNormalizedString` -> `Parse` then compare all properties

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnresolvedMessages.cs (hunk 1 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnresolvedMessages.cs]: Aren't the tests the truth table? :) 
Feels like I'd be repeating the tests.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnresolvedMessages.cs]: The tests have comments that go in that much detail.

File: test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionRangeTests.cs (hunk 2 above)
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionRangeTests.cs:179]: Good point. I thought there was one, but I couldn't find it. 

I'll add a few.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnresolvedMessages.cs (hunk 1 above)
Comment (joelverhagen) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnresolvedMessages.cs]: I've found that big code comments in the product code -- more of the "why" than the "how" -- help explain code that could look weird to a new reader. I feel like the reality is that test code does not get the same kind of understanding/process/quality checks as product code. In particular if someone is confused about some behavior it's pretty easy to find the product code on GitHub but less easy to find the related tests to explain the behavior. Just a suggestion though -- your call.

File: test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionRangeTests.cs (hunk 2 above)
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionRangeTests.cs:179]: Fixed.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnresolvedMessages.cs (hunk 1 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnresolvedMessages.cs]: I'll look into it. 

I just don't want to be too verbose ;)

**Comment by nkolev92** [comment]
@joelverhagen Added a truth table, ready for another look.

@donnie-msft Let me if you think other examples are needed.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnresolvedMessages.cs (hunk 1 above)
Comment (donnie-msft) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnresolvedMessages.cs]: > @joelverhagen Added a truth table, ready for another look.
> 
> @donnie-msft Let me if you think other examples are needed.

Looks good. I'm not a fan of the truth table in the code since you are already showing examples in the tests. But I understand why some would prefer to see it here.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnresolvedMessages.cs [hunk 3]
```diff
@@ -263,10 +263,20 @@ internal static NuGetVersion GetBestMatch(SortedSet<NuGetVersion> versions, Vers
...
                 }
             }
-
-            if (!range.IsFloating || range.HasLowerBound || range.HasUpperBound)
+            //| Range          | Available    | Closest |
+            //| [1.0.0, )      | 0.7.0, 0.9.0               | 0.7.0    |
+            //| (0.5.0, 1.0.0) | 0.1.0, 1.0.0               | 1.0.0    |
+            //| (, 1.0.0)      | 2.0.0, 3.0.0               | 2.0.0    |
+            //| [1.*,)         | 0.0.1, 0.0.5, 0.1.0, 0.9.0 | 0.9.0    |
+            //| [1.*, 2.0.0]   | 0.1.0,0.3.0,3.0.0,4.0.0    | 3.0.0    |
```
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnresolvedMessages.cs]: nit: spaces around comma

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnresolvedMessages.cs [hunk 4]
```diff
@@ -263,10 +263,20 @@ internal static NuGetVersion GetBestMatch(SortedSet<NuGetVersion> versions, Vers
                     ideal = range.MinVersion;
                 }
             }
-
-            if (!range.IsFloating || range.HasLowerBound || range.HasUpperBound)
+            //| Range          | Available    | Closest |
```
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnresolvedMessages.cs]: nit: align column heading with a blast of spaces

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs [hunk 5]
```diff
@@ -9045,8 +9045,117 @@ public void RestoreNetCore_PackagesLockFile_WithProjectChangeFramweorksAndLocked
             }
         }
 
-        
+                [Theory]
```
Comment (joelverhagen) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs]: nit extra indentation

**Review by joelverhagen — APPROVED** [review]
Consider mitigating risk by choosing a strategic release to land this in /cc @rrelyea

**Review by zivkan — APPROVED** [review]

**Comment by nkolev92** [comment]
Discussed with @rrelyea. 
Decided to take this in 5.5, so this PR will stay open until next week when dev switches to 5.5.
